binderhub/app.py

```diff
@@ -144,7 +144,8 @@ def initialize(self, *args, **kwargs):
         self.tornado_app = tornado.web.Application([
             (r"/build/([a-z0-9]+)/([^?]+)", BuildHandler),
             (r"/redirect", RedirectHandler),
-            (r"/", MainHandler)
+            (r"/v2/([a-z0-9]+)/([^?]+)", ParameterizedMainHandler),
```

Comment:
We can greedily match any url part, so that ParameterizedMain gets called and can handle a more informative "no such provider" error. Plus, this pattern only applies to path, so there's no need to split on `?`, e.g. `v2/([^/])/(.*)`